packages/core/src/utils/cursor-pagination/helpers.ts

```diff
@@ -0,0 +1,82 @@
+import { type SelectQueryBuilder, sql } from "kysely"
+import type { Cursor, Pageable } from "./types"
+
+type GetNextCursorOptions<T> = {
+  pageable: Pageable
+  buildCursor: (record: T) => Cursor
+}
+
+function getNextCursor<T>(records: T[], options: GetNextCursorOptions<T>): Cursor | null {
+  // We fetched take+1 records to determine if there is a record behind `take`.
+  // If we got exactly `take+1` records, there is a next page.
+  const hasNextPage = records.length === options.pageable.take + 1
+  if (!hasNextPage) {
+    return null
+  }
+
+  // biome-ignore lint/style/noNonNullAssertion: We know there is a record at -1 from the check above.
+  return options.buildCursor(records.at(-1)!)
+}
+
+type PaginatedQueryOptions<T> = {
+  pageable: Pageable
+  decodeCursor: (cursor: Cursor) => string
+  buildCursor: (record: T) => Cursor
+  column: string
+  order: "asc" | "desc"
+}
+
+/**
+ * Add cursor based pagination to a query based on ordering by a single column, e.g. created_at.
+ *
+ * Cursors are opaque strings that represent the next record that will be fetched in the ordered set of records based on the given column.
+ * A cursor value of `null` will return no records, and is returned when there are no extra records to fetch.
+ * A cursor value of `undefined` will return the first `take` records.
+ *
+ * For more complex queries that require ordering by multiple columns, manual implementation
+ * is needed.
+ */
+export async function singleColPaginatedQuery<DB, TB extends keyof DB, T>(
```

Comment:
Can't we just support `columns = ["id", "X"]`, and paginate accordingly?

```suggestion
export async function PaginatedQuery<DB, TB extends keyof DB, T>(
```

Reply:
Then what will happen to the where clause? Can you make this generic to multiple columns?

Reply:
I think you can just use `where (columnA, columnB) < (X, Y)` https://use-the-index-luke.com/sql/partial-results/fetch-next-page

Reply:
Should be very much possible, just need to rearrange the sort orders too

Reply:
Alright, I'll take a look at it